crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2029,40 +2060,85 @@ impl AuthorityPerEpochStore {
+                sequenced_transactions.push(tx);
+            }
+        }
         let mut batch = self.db_batch();
 
-        sequenced_transactions.extend(
-            self.load_deferred_transactions_for_consensus_round(&mut batch, commit_round)?
-                .into_iter(),
-        );
+        // Pre-process transactions to find the most recent randomness round included in the commit.
+        let mut last_randomness_round_written = self.last_randomness_round_written()?;
+        // There must be at most one RandomnessStateUpdate per commit.
+        let mut randomness_state_update_found = None;
+        for tx in system_transactions.iter() {
+            let SequencedConsensusTransactionKind::System(tx) = &tx.0.transaction else {
+                unreachable!("system_transactions vector should only contain system transactions")
+            };
+            if let TransactionKind::RandomnessStateUpdate(rsu) =
+                tx.data().intent_message().value.kind()
+            {
+                assert!(
+                    randomness_state_update_found.is_none(),
+                    "found multiple RandomnessStateUpdates in one commit: {:?}, {rsu:?}",
+                    randomness_state_update_found.unwrap(),
+                );
+                randomness_state_update_found = Some(rsu);
+                last_randomness_round_written = std::cmp::max(
+                    last_randomness_round_written,
+                    RandomnessRound(rsu.randomness_round),
+                );
+            }
+        }
 
-        // TODO: This is a no-op until we start using random round transactions
-        let placeholder_random_round = u64::MAX - 1;
-        sequenced_transactions.extend(
-            self.load_deferred_transactions_for_randomness_round(
+        // Load transactions deferred from prevous commits.
+        // We do this after updating the last_randomness_round_written above so that every deferred
+        // transaction that can be run with this commit is loaded.
+        let deferred_tx: Vec<VerifiedSequencedConsensusTransaction> = self
+            .load_deferred_transactions_for_consensus_round(&mut batch, commit_round)?
+            .into_iter()
+            .chain(self.load_deferred_transactions_for_randomness_round(
                 &mut batch,
-                placeholder_random_round,
-            )?
-            .into_iter(),
-        );
+                last_randomness_round_written,
+            )?)
+            .collect();
+        let previously_deferred_tx_digests: HashSet<_> = deferred_tx
+            .iter()
+            .map(|tx| match tx.0.transaction.key() {
+                SequencedConsensusTransactionKey::External(
+                    ConsensusTransactionKey::Certificate(digest),
+                ) => digest,
+                _ => panic!("deferred transaction was not a user certificate: {tx:?}"),
+            })
+            .collect();
+        sequenced_transactions.extend(deferred_tx.into_iter());
```

Comment:
Is it now possible to have a duplicated transaction? one that was deferred earlier, and has now re-appeared in the current commit?

Reply:
`self.record_consensus_message_processed(batch, key.clone())?;` is called for deferred tx so they will still be de-duped